test/DiTest.php

```diff
@@ -1017,4 +1025,102 @@ public function testGetInstanceWithParamsHasSameNameAsDependencyParam()
+            ],
+        ];
+        $di = new Di(null, null, new Config($config));
+        $ref = new \ReflectionObject($di);
+        $method = $ref->getMethod('resolveMethodParameters');
+        $method->setAccessible(true);
+
+        $args = [
+            'ZendTest\\Di\\TestAsset\\ConstructorInjection\\B',
+            '__construct',
+            [],
+            null,
+            Di::METHOD_IS_CONSTRUCTOR,
+            true
+        ];
+        $res = $method->invokeArgs($di, $args);
+        $this->assertInstanceOf('ZendTest\\Di\\TestAsset\\ConstructorInjection\\E', $res[0], 'the first resolved parameter for constructor type preferenced');
+
+        //user provides params
+        $args = [
+            'ZendTest\\Di\\TestAsset\\ConstructorInjection\\B',
+            '__construct',
+            ['a' => 'ZendTest\\Di\\TestAsset\\ConstructorInjection\\F'],
+            null,
+            Di::METHOD_IS_CONSTRUCTOR,
+            true
+        ];
+        $res = $method->invokeArgs($di, $args);
+        $this->assertInstanceOf('ZendTest\\Di\\TestAsset\\ConstructorInjection\\F', $res[0], 'the first resolved parameter user provided ');
+
+        //user provides alias name
+        $im = $di->instanceManager();
+        $im->addAlias('foo', 'ZendTest\\Di\\TestAsset\\ConstructorInjection\\F', ['params' => ['p' => 'v1']]);
+        $im->addAlias('bar', 'ZendTest\\Di\\TestAsset\\ConstructorInjection\\F', ['params' => ['p' => 'v2']]);
+
+        $args = [
+            'ZendTest\\Di\\TestAsset\\ConstructorInjection\\B',
+            '__construct',
+            ['a' => 'foo'],
+            null,
+            Di::METHOD_IS_CONSTRUCTOR,
+            true
+        ];
+        $res = $method->invokeArgs($di, $args);
+        $this->assertInstanceOf('ZendTest\\Di\\TestAsset\\ConstructorInjection\\F', $res[0], 'the first resolved parameter user provided with alias parameter');
+        $this->assertEquals('v1', $res[0]->params['p']);
+
+        $args = [
+            'ZendTest\\Di\\TestAsset\\ConstructorInjection\\B',
+            '__construct',
+            ['a' => 'bar'],
+            null,
+            Di::METHOD_IS_CONSTRUCTOR,
+            true
+        ];
+        $res = $method->invokeArgs($di, $args);
+        $this->assertInstanceOf('ZendTest\\Di\\TestAsset\\ConstructorInjection\\F', $res[0], 'the first resolved parameter user provided with alias parameter');
+        $this->assertEquals('v2', $res[0]->params['p']);
+    }
+
+    public function testAliasesWithDeffrentParams()
+    {
+        $config = [
+            'instance' => [
+                'preference' => [
+                    'ZendTest\\Di\\TestAsset\\ConstructorInjection\\A' => 'ZendTest\\Di\\TestAsset\\ConstructorInjection\\E',
+                ],
+            ],
+        ];
+        $di = new Di(null, null, new Config($config));
+        $im = $di->instanceManager();
+        $im->addAlias('foo', 'ZendTest\\Di\\TestAsset\\ConstructorInjection\\F', ['params' => ['p' => 'vfoo']]);
```

Comment:
```php
'ZendTest\\Di\\TestAsset\\ConstructorInjection\\F'
```
as of PHP 5.5 can be written as

```php
\ZendTest\Di\TestAsset\ConstructorInjection\F::class
```

This makes it a class reference, rather than a string reference ;-)